From 96cb7819513f72926e215c63b22931a50fc7aab2 Mon Sep 17 00:00:00 2001
From: EternalMikaelson <klaus_mikaelson@rediffmail.com>
Date: Fri, 9 Oct 2026 01:01:17 +0000
Subject: [PATCH 5/5] gallery: Fix broken 'Set as wallpaper'

WallpaperManager.getCropAndSetWallpaperIntent(Uri) re-derives the
URI's MIME type via ContentResolver.getType() and requires it to
start with 'image/'. MediaProvider fails that check for at least
some MediaStore images on this ROM, throwing IllegalArgumentException
('Cannot use passed URI to set wallpaper'), silently caught by
PictureEditorDM.a() - so 'Set as wallpaper' does nothing.

Replace with ACTION_ATTACH_DATA + explicit setDataAndType(uri,
'image/*'): declares the MIME type directly on the Intent instead of
relying on the ContentProvider's own type resolution, so the
wallpaper-crop activity never hits the failing check.

Change-Id: Ib6507e5dd5a0a480f58686b43b437030c0d2a22c
Signed-off-by: EternalMikaelson <klaus_mikaelson@rediffmail.com>
---
 .../pictureeditorpage/PictureEditorDM.smali   | 139 ++----------------
 1 file changed, 11 insertions(+), 128 deletions(-)

diff --git a/smali_classes13/com/oplus/gallery/pictureeditorpage/PictureEditorDM.smali b/smali_classes13/com/oplus/gallery/pictureeditorpage/PictureEditorDM.smali
index 771b6a49b..b7ed0f76f 100644
--- a/smali_classes13/com/oplus/gallery/pictureeditorpage/PictureEditorDM.smali
+++ b/smali_classes13/com/oplus/gallery/pictureeditorpage/PictureEditorDM.smali
@@ -43,153 +43,36 @@
 
 # virtual methods
 .method public final a(Landroid/app/Activity;Landroid/net/Uri;)V
-    .locals 4
+    .locals 3
 
-    .line 1
     const-string p0, "activity"
 
-    .line 2
-    .line 3
     invoke-static {p1, p0}, Lkotlin/jvm/internal/Intrinsics;->checkNotNullParameter(Ljava/lang/Object;Ljava/lang/String;)V
 
-    .line 4
-    .line 5
-    .line 6
     const-string p0, "pickedItem"
 
-    .line 7
-    .line 8
     invoke-static {p2, p0}, Lkotlin/jvm/internal/Intrinsics;->checkNotNullParameter(Ljava/lang/Object;Ljava/lang/String;)V
 
-    .line 9
-    .line 10
-    .line 11
-    const-string p0, "WallpaperSetUtil"
+    new-instance v0, Landroid/content/Intent;
 
-    .line 12
-    .line 13
-    const/4 v0, 0x0
+    const-string v1, "android.intent.action.ATTACH_DATA"
 
-    .line 14
-    :try_start_0
-    invoke-static {p1}, Landroid/app/WallpaperManager;->getInstance(Landroid/content/Context;)Landroid/app/WallpaperManager;
+    invoke-direct {v0, v1}, Landroid/content/Intent;-><init>(Ljava/lang/String;)V
 
-    .line 15
-    .line 16
-    .line 17
-    move-result-object v1
+    const-string v1, "image/*"
 
-    .line 18
-    if-nez v1, :cond_0
-
-    .line 19
-    .line 20
-    const-string p1, "setWallPaper, mWallpaperManager is null"
+    invoke-virtual {v0, p2, v1}, Landroid/content/Intent;->setDataAndType(Landroid/net/Uri;Ljava/lang/String;)Landroid/content/Intent;
 
-    .line 21
-    .line 22
-    invoke-static {p0, p1, v0}, Lcom/oplus/aiunit/vision/b3a;->q(Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
+    const/4 v2, 0x1
 
-    .line 23
-    .line 24
-    .line 25
-    goto :goto_1
+    invoke-virtual {v0, v2}, Landroid/content/Intent;->setFlags(I)Landroid/content/Intent;
 
-    .line 26
-    :catch_0
-    move-exception p1
+    const-string v2, "mimeType"
 
-    .line 27
-    goto :goto_0
+    invoke-virtual {v0, v2, v1}, Landroid/content/Intent;->putExtra(Ljava/lang/String;Ljava/lang/String;)Landroid/content/Intent;
 
-    .line 28
-    :cond_0
-    invoke-virtual {v1, p2}, Landroid/app/WallpaperManager;->getCropAndSetWallpaperIntent(Landroid/net/Uri;)Landroid/content/Intent;
+    invoke-virtual {p1, v0}, Landroid/app/Activity;->startActivity(Landroid/content/Intent;)V
 
-    .line 29
-    .line 30
-    .line 31
-    move-result-object v1
-
-    .line 32
-    const-string v2, "from"
-
-    .line 33
-    .line 34
-    sget-object v3, Lcom/oplus/aiunit/vision/hw5;->a:Landroid/content/Context;
-
-    .line 35
-    .line 36
-    invoke-virtual {v3}, Landroid/content/Context;->getPackageName()Ljava/lang/String;
-
-    .line 37
-    .line 38
-    .line 39
-    move-result-object v3
-
-    .line 40
-    invoke-virtual {v1, v2, v3}, Landroid/content/Intent;->putExtra(Ljava/lang/String;Ljava/lang/String;)Landroid/content/Intent;
-
-    .line 41
-    .line 42
-    .line 43
-    invoke-virtual {p1, v1}, Landroid/app/Activity;->startActivity(Landroid/content/Intent;)V
-    :try_end_0
-    .catch Ljava/lang/Exception; {:try_start_0 .. :try_end_0} :catch_0
-
-    .line 44
-    .line 45
-    .line 46
-    goto :goto_1
-
-    .line 47
-    :goto_0
-    new-instance v1, Ljava/lang/StringBuilder;
-
-    .line 48
-    .line 49
-    const-string v2, "setWallpaperSystem:exception = "
-
-    .line 50
-    .line 51
-    invoke-direct {v1, v2}, Ljava/lang/StringBuilder;-><init>(Ljava/lang/String;)V
-
-    .line 52
-    .line 53
-    .line 54
-    invoke-virtual {v1, p1}, Ljava/lang/StringBuilder;->append(Ljava/lang/Object;)Ljava/lang/StringBuilder;
-
-    .line 55
-    .line 56
-    .line 57
-    const-string p1, " uri: "
-
-    .line 58
-    .line 59
-    invoke-virtual {v1, p1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
-
-    .line 60
-    .line 61
-    .line 62
-    invoke-virtual {v1, p2}, Ljava/lang/StringBuilder;->append(Ljava/lang/Object;)Ljava/lang/StringBuilder;
-
-    .line 63
-    .line 64
-    .line 65
-    invoke-virtual {v1}, Ljava/lang/StringBuilder;->toString()Ljava/lang/String;
-
-    .line 66
-    .line 67
-    .line 68
-    move-result-object p1
-
-    .line 69
-    invoke-static {p0, p1, v0}, Lcom/oplus/aiunit/vision/b3a;->q(Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
-
-    .line 70
-    .line 71
-    .line 72
-    :goto_1
     return-void
 .end method
 
-- 
2.43.0

